model-optimizer/mo/front/common/partial_infer/utils.py

```diff
@@ -6,6 +6,158 @@
+    Compare if dim1 is equal to dim2 or any of them is dynamic
+
+    :param dim1: dimension to compare
+    :param dim2: dimension to compare
+    :return: boolean result of the comparison
+    """
+    return dim1 is dynamic_dimension or dim2 is dynamic_dimension or dim1 == dim2
+
+
+def compare_shapes(shape1, shape2):
+    """
+    Compares two shape tensors. The shapes are considered equal if they have the same rank and the corresponding
+    dimensions are either equal or at least one of them is dynamic.
+
+    :param shape1: the first shape to compare
+    :param shape2: the second shape to compare
+    :return: boolean result of the comparison
+    """
+    if shape1.ndim != shape2.ndim:
+        return False
+    if shape1.size != shape2.size:
+        return False
+    for d1, d2 in zip(shape1, shape2):
+        if not compare_dimensions(d1, d2):
+            return False
+    return True
+
+
+def strict_compare_tensors(tensor1, tensor2):
+    """
+    Strict comparison of two tensors. The tensors are equal iff their corresponding elements are equal or both are
+    dynamic.
+
+    :param tensor1: the first tensor to compare
+    :param tensor2: the second tensor to compare
+    :return: boolean result of the comparison
+    """
+    if not isinstance(tensor1, np.ma.masked_array):
+        tensor1 = shape_array(tensor1)
+    if not isinstance(tensor2, np.ma.masked_array):
+        tensor2 = shape_array(tensor2)
+
+    if tensor1.ndim != tensor2.ndim:
+        return False
+    if tensor1.size != tensor2.size:
+        return False
+    if tensor1.ndim == 0:
+        return tensor1.item() == tensor2.item()
+    if not np.array_equal(tensor1.shape, tensor2.shape):
+        return False
+    for d1, d2 in zip(tensor1.flatten(), tensor2.flatten()):
+        if d1 is not dynamic_dimension and d2 is not dynamic_dimension:
+            if d1 != d2:
+                return False
+        elif d1 is not dynamic_dimension or d2 is not dynamic_dimension:
+            return False
+    return True
+
+
+def shape_delete(shape: np.ma.masked_array, obj: [int, list]):
+    """
+    Removes element in the input tensor shape (presumably the numpy masked array) specified by index/indices obj.
+    The function is implemented to avoid usage of np.delete which corrupts information about the masked elements.
+
+    :param shape: the shape object to remove elements from
+    :param obj: the list or a single integer defining index(es) of elements to remove
+    :return: shape with removed selected elements
+    """
+    if isinstance(obj, (int, np.int64, np.int32)):
+        return shape_delete(shape, [obj])
+    elif isinstance(obj, np.ndarray):
+        return shape_delete(shape, obj.tolist())
+    elif isinstance(obj, list):
+        result = shape.copy()
+        obj = [item if item >= 0 else len(shape) + item for item in obj]
+        for index in sorted(obj, reverse=True):
+            result = np.ma.concatenate((result[:index], result[index + 1:]))
```

Comment:
pls add assert before concatenation:
`assert 0 <= index and index < len(result)`
 
I think that something bad will happen if `index` is in one of the corners.

Reply:
Changed.